target/linux/generic/pending-6.6/730-net-phy-realtek-detect-early-version-of-RTL8221B.patch

```diff
@@ -10,37 +10,26 @@ Implement custom identify function using the PKGID instead of iterating
 over the implemented MMDs.
 
 Signed-off-by: Daniel Golle <daniel@makrotopia.org>
----
- drivers/net/phy/realtek.c | 50 ++++++++++++++++++++++++++++++++++++++-
- 1 file changed, 49 insertions(+), 1 deletion(-)
 
 --- a/drivers/net/phy/realtek.c
 +++ b/drivers/net/phy/realtek.c
 @@ -81,6 +81,7 @@
  
  #define RTL_GENERIC_PHYID			0x001cc800
  #define RTL_8211FVD_PHYID			0x001cc878
-+#define RTL_8221B_VB_CG				0x001cc849
++#define RTL_8221B_VB_CG			0x001cc849
  
  MODULE_DESCRIPTION("Realtek PHY driver");
  MODULE_AUTHOR("Johnson Leung");
-@@ -801,6 +802,54 @@ static int rtl822x_probe(struct phy_devi
- 	return 0;
+@@ -782,6 +783,38 @@ static int rtl8226_match_phy_device(stru
+ 	       rtlgen_supports_2_5gbps(phydev);
  }
  
 +static int rtl8221b_vb_cg_match_phy_device(struct phy_device *phydev)
 +{
 +	int val;
 +	u32 id;
 +
-+	if (phydev->is_c45) {
-+		if (phydev->c45_ids.device_ids[1])
-+			return phydev->c45_ids.device_ids[1] == RTL_8221B_VB_CG;
-+	} else {
-+		if (phydev->phy_id)
-+			return phydev->phy_id == RTL_8221B_VB_CG;
-+	}
```

Comment:
To me, this looks like an optimization that skips having to talk with the phy.
Having it is probably worthwhile.

If it is like dango writes in the patch description, this early check should only return if the id from the scan is invalid (`0` if `!is_c45`).
@RolandoMagico checked that the 'invalid' id is `0xFFFFFFFF` when `is_c45`.

I guess, it would be better to check for `phydev->c45_ids.device_ids[1] != 0xFFFFFFFF` rather than removing the entire early check.

Reply:
The commit https://github.com/openwrt/openwrt/commit/f6c27b22f6f84b13ad230dcf3afa7f11efe7a8bc was probably created because after upgrading to 6.6 kernel, a restart loop was detected during boot. The restart issue was resolved, but there was an issue with Realtek PHY detection.
After my analysis, it turned out that the reason for the restart was the change https://github.com/openwrt/openwrt/commit/5b6bca812dee241eaa4aa13aed295926550b59c9#diff-0c1ba991ce1092fc7697b97481da32349ccaaac650ad0a454e2d 9 9a 6d1bfe1f8. The line "if (phydev->drv->read_mmd) made a startup error.
This PR restores recent changes and replaces "if (phydev->drv->read_mmd) {" with "if (phydev->mdio.bus->read_c45) {", which resolves all previous issues.

Reply:
See info below